migrate/migrate.go

```diff
@@ -13,55 +20,213 @@ import (
 	dshelp "github.com/ipfs/go-ipfs-ds-help"
 )
 
+const (
+	PagesBeforeSleep = 5 // Up to 5000 entries given a default page size of 1000
+	PaginationSleep  = 1 * time.Second
+
+	RequestTimeout    = 3 * time.Second
+	RetryDelay        = 250 * time.Millisecond
+	NumRequestRetries = 3
+)
+
+var (
+	// Use a concurrent map to keep track of the pin status of CIDs read from S3
+	pinMap = sync.Map{}
+
+	// Default context
+	ctx = context.Background()
+)
+
 func Migrate(bucket, prefix, ipfsUrl string) {
+	// Setup S3 access and pagination
 	cfg := configureAWS()
 	s3Client := s3.NewFromConfig(cfg)
-
-	output, err := s3Client.ListObjectsV2(context.TODO(), &s3.ListObjectsV2Input{
-		Bucket:  aws.String(bucket),
-		Prefix:  aws.String(prefix),
-		MaxKeys: 100,
+	paginator := s3.NewListObjectsV2Paginator(s3Client, &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
 	})
 
+	// Use a single shell to IPFS since it uses a thread-safe HTTP client
+	ipfsShell := shell.NewShell(ipfsUrl)
+
+	// If statuses were already written previously, load them, and pin any unpinned CIDs.
+	_, unpinnedCids := readPinStatuses()
+	pinCids(ipfsShell, unpinnedCids)
+
+	// Write final statuses at the end
+	defer writePinStatuses()
+
+	pageNum := 1
+	for paginator.HasMorePages() {
+		// Pagination is stateful in order to keep track of position and so should not be placed in a goroutine
+		page, err := nextPage(paginator)
+		if err != nil { // We've retried and still failed, exit the loop.
+			log.Printf("page error: %s", err)
+			break
+		}
+		log.Printf("retrieved page: %d", pageNum)
+		pinCids(ipfsShell, cidsFromPage(page))
+		if (pageNum % PagesBeforeSleep) == 0 {
+			// Sleep before pulling more pages to avoid S3 throttling
+			time.Sleep(PaginationSleep)
+		}
+		pageNum++
+	}
+}
+
+func configureAWS() aws.Config {
+	cfg, err := config.LoadDefaultConfig(ctx)
 	if err != nil {
 		log.Fatal(err)
 	}
+	return cfg
+}
 
-	sh := shell.NewShell(ipfsUrl)
+func blockToCid(key string) (string, error) {
+	// All files will be either in `dag-jose` (multicodec 0x85) or `dag-cbor` (multicodec 0x71) format. Since `dag-cbor`
+	// is a superset of `dag-jose`, use `dag-cbor` for the migration.
+	cid, err := dshelp.DsKeyToCidV1(ds.NewKey(key), 0x71)
+	return cid.String(), err
+}
+
+func nextPage(paginator *s3.ListObjectsV2Paginator) (*s3.ListObjectsV2Output, error) {
+	for i := 0; i < NumRequestRetries; i++ {
+		page, err := func() (*s3.ListObjectsV2Output, error) {
```

Comment:
why are we creating this function inside the loop?

Reply:
Making the function was necessary to make sure the `defer cancel()` [worked correctly](https://stackoverflow.com/questions/45617758/defer-in-the-loop-what-will-be-better).

Making it an anonymous function versus a named function was to avoid proliferating functions too much - I didn't like the idea of a `pinCidWithRetries` function calling a separate `pinCid` function (similarly for the pagination flow).

I can do it the latter way though (that is how it was earlier until I merged the two).